src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,99 @@
   }
 
-  /**
-   * For testing purpose only.
-   * @param map 
-   */
-  ClientSideCache(Map<ByteBuffer, Object> map) {
-    this.cache = map;
-  }
+  public abstract void invalidateAll();
 
-  public final void clear() {
-    cache.clear();
-  }
+  protected abstract void invalidateAll(Iterable<Long> hashes);
 
-  public final void invalidateKeys(List list) {
+  final void invalidate(List list) {
     if (list == null) {
-      clear();
+      invalidateAll();
       return;
     }
 
-    list.forEach(this::invalidateKey);
+    list.forEach(this::invalidate0);
   }
 
-  private void invalidateKey(Object key) {
-    if (key instanceof byte[]) {
-      cache.remove(convertKey((byte[]) key));
-    } else {
-      throw new JedisException("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
+  private void invalidate0(Object key) {
+    if (!(key instanceof byte[])) {
+      throw new AssertionError("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
     }
-  }
 
-  protected void setKey(Object key, Object value) {
-    cache.put(getMapKey(key), value);
-  }
+    final ByteBuffer mapKey = makeKey((byte[]) key);
 
-  protected <T> T getValue(Object key) {
-    return (T) getMapValue(key);
+    Set<Long> hashes = keyHashes.get(mapKey);
+    if (hashes != null) {
+      writeLock.lock();
+      try {
+        invalidateAll(hashes);
+        keyHashes.remove(mapKey);
+      } finally {
+        writeLock.unlock();
+      }
+    }
   }
 
-  private Object getMapValue(Object key) {
-    return cache.get(getMapKey(key));
-  }
+  protected abstract void put(long hash, Object value);
+
+  protected abstract Object get(long hash);
 
-  private ByteBuffer getMapKey(Object key) {
-    if (key instanceof byte[]) {
-      return convertKey((byte[]) key);
-    } else {
-      return convertKey(SafeEncoder.encode(String.valueOf(key)));
+  final <T> T getValue(Function<CommandObject<T>, T> loader, CommandObject<T> command, String... keys) {
+
+    final long hash = getHash(command);
+
+    T value = (T) get(hash);
+    if (value != null) {
+      return value;
+    }
+
+    value = loader.apply(command);
+    if (value != null) {
+      writeLock.lock();
+      try {
+        put(hash, value);
+        for (String key : keys) {
+          ByteBuffer mapKey = makeKey(key);
+          if (keyHashes.containsKey(mapKey)) {
+            keyHashes.get(mapKey).add(hash);
+          } else {
+            Set<Long> set = new HashSet<>();
+            set.add(hash);
+            keyHashes.put(mapKey, set);
+          }
+        }
+      } finally {
+        writeLock.unlock();
+      }
     }
+
+    return value;
+  }
+
+  protected abstract long getHash(CommandObject command);
+
+  private ByteBuffer makeKey(String key) {
+    return makeKey(SafeEncoder.encode(key));
   }
 
-  private static ByteBuffer convertKey(byte[] b) {
+  private static ByteBuffer makeKey(byte[] b) {
```

Comment:
Github ate my comment :(

WDYT about make instead of `makeKey`? Or, were you thinking that there would be a different type of make (makeEntry, makeCommandKey, etc)?

Reply:
Here we are creating a key for our key to command hashes map. Only `make` would be worse.